### PR TITLE
Add support biblatex extended name format

### LINKED ISSUE #4558: Wrong sorting by author/editor when BibTeX extended name format is used
JabRef version 4.3.1

- [ ] I have tested the latest development version from http://builds.jabref.org/master/ and the problem persists

1. Have an entry in the database in which the first author name is written using the BibTeX extended name format, for example:
```
@InProceedings{Hasselt2016,
  author     = {family=Hasselt, given=Hado P., prefix=van useprefix=false and Guez, Arthur and Hessel, Matteo and Mnih, Volodymyr and Silver, David},
  title      = {Learning Values across Many Orders of Magnitude},
  booktitle  = {Advances in Neural Information Processing Systems},
  date       = {2016},
  pages      = {4287--4295},
  url        = {http://papers.nips.cc/paper/6076-learning-values-across-many-orders-of-magnitude},
  urldate    = {2017-06-07}
}
```
2. In the main table, click the header of the column _author/editor_ to sort entries by author/editor name.

Expected behavior: the example entry should be listed under the letter H (for Hasselt) or maybe V (for the van prefix).

Actual behavior: The example entry is listed first in my database, probably using the "f" letter of the "family" word for sorting.

Additional info: The BibTeX extended name format is somewhat described in section 3.8 of the [Biber manual](http://mirrors.ctan.org/biblio/biber/documentation/biber.pdf).

### PR DESCRIPTION
**Pull request by u7302654**
Description:
This pull request adds support for parsing BibLaTeX extended name format, which includes fields including family, given, prefix, suffix, and useprefix.  This feature ensures correct parsing and formatting of author names with complex structures, including options to handle prefixes such as "van" and to manage cases where the prefix should be ignored (useprefix=false).

These changes solve the problem described in [Issue #4558](https://github.com/JabRef/jabref/issues/4558), where entries using the extended BibLaTeX format were not being parsed and sorted correctly, leading to incorrect name sorting behaviour.

Changes:
- Added a new method parseExtendedNameFormat in AuthorListParser to handle the extended name format.
- Created tests to ensure correct behaviour for author names with different combinations of family, given, prefix, and useprefix values.
- Modified the author list parsing logic to include these new parsing rules.

Closes #4558

### Mandatory checks

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [/] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [/] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by koppor** [comment]
To ease organizational workflows I have linked the pull-request to the issue with syntax as described in https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

> <h2 id="linking-a-pull-request-to-an-issue-using-a-keyword">Linking a pull request to an issue using a keyword</h2>
> <p>You can link a pull request to an issue by using a supported keyword 
> in the pull request's description or in a commit message. The pull 
> request <strong>must be</strong> on the default branch.</p>
> <ul>
> <li>close</li>
> <li>closes</li>
> <li>closed</li>
> <li>fix</li>
> <li>fixes</li>
> <li>fixed</li>
> <li>resolve</li>
> <li>resolves</li>
> <li>resolved</li>
> </ul>
> <p>If you use a keyword to reference a pull request comment in another 
> pull request, the pull requests will be linked. Merging the referencing 
> pull request also closes the referenced pull request.</p>
> <p>The syntax for closing keywords depends on whether the issue is in the same repository as the pull request.</p>


### Example:
- ✅ `Fixes #xyz` links pull-request to issue. Merging the PR will close the Issue.
- ✅ `Fixes https://github.com/JabRef/jabref/issues/xyz` links pull-request to issue. Merging the PR will close the Issue.
- ✅ `Fixes https://github.com/Koppor/jabref/issues/xyz` links pull-request to issue. Merging the PR will close the Issue.
- ❌ `Fixes [#xyz](https://github.com/JabRef/jabref/issues/xyz)` links pull-request to issue. Merging the PR will **NOT** close the Issue.

**Comment by koppor** [comment]
I did not see any JavaFX component implementing a new JavaFX component for the entry editor: https://github.com/JabRef/jabref/issues/4558#issuecomment-1904127621. It is OK to focus on the logic only. It is OK, but then we need to create a follow-up issue.

@u7302654 Please fix checkstyle issues. You seem to have missed [Step 3: Set up JabRef's code style](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html).

Moreover, please add a screenshot to show that the sorting issue of the original poster is solved (https://github.com/JabRef/jabref/issues/4558#issue-394698254). If it is not solved, it is also OK, because it improves the data model, but then we cannot close the issue at all.

File: src/main/java/org/jabref/logic/importer/AuthorListParser.java [hunk 1]
```diff
@@ -147,59 +140,37 @@ public AuthorList parse(@NonNull String listOfNames) {
         listOfNames = simpleNormalForm.authors;
         boolean andOthersPresent = simpleNormalForm.andOthersPresent;
 
-        // Handle case names in order lastname, firstname and separated by ","
-        // E.g., Ali Babar, M., Dingsøyr, T., Lago, P., van der Vliet, H.
-        final boolean authorsContainAND = listOfNames.toUpperCase(Locale.ENGLISH).contains(" AND ");
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/AuthorListParser.java:152]: This code was removed - therefore `org.jabref.logic.formatter.bibtexfields.NormalizeNamesFormatterTest` is failing. Please fix. 😅

**Review by koppor — CHANGES_REQUESTED** [review]

File: src/main/java/org/jabref/logic/importer/AuthorListParser.java [hunk 2]
```diff
@@ -209,6 +180,100 @@ public AuthorList parse(@NonNull String listOfNames) {
...
+     * Attempts to parse a single author string using the BibLaTeX extended name format.
+     * The format includes attributes such as family, given, prefix, and suffix.
+     * Also handles the 'useprefix=false' case where the prefix should be ignored.
+     *
+     * @param authorString the string representing a single author in extended format
+     * @return Optional containing an Author object if parsing is successful, or empty if not
+     */
+    private Optional<Author> parseExtendedNameFormat(String authorString) {
+        Map<String, String> nameParts = new HashMap<>();
+        Matcher matcher = Pattern.compile("(\\w+)\\s*=\\s*([^,]+)(?:,\\s*|$)").matcher(authorString);
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/logic/importer/AuthorListParser.java:193]: Please compile the regex only once. Compiling regular expressions is an expensive operation.

File: src/main/java/org/jabref/logic/importer/AuthorListParser.java [hunk 3]
```diff
@@ -209,6 +180,100 @@ public AuthorList parse(@NonNull String listOfNames) {
...
+        while (matcher.find()) {
+            nameParts.put(matcher.group(1).trim(), matcher.group(2).trim());
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/logic/importer/AuthorListParser.java:195]: nitpick: Named groups would make this code clearer

File: src/main/java/org/jabref/logic/importer/AuthorListParser.java [hunk 4]
```diff
@@ -209,6 +180,100 @@ public AuthorList parse(@NonNull String listOfNames) {
...
+        int start = 0;
+        for (int i = 0; i < authorList.length(); i++) {
+            char c = authorList.charAt(i);
+            if (c == '{') {
+                bracesLevel++;
+            } else if (c == '}') {
+                bracesLevel--;
+            } else if (i <= authorList.length() - 5 && authorList.substring(i, i + 5).equals(" and ") && bracesLevel == 0) {
+                authors.add(authorList.substring(start, i));
+                i += 4;
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/logic/importer/AuthorListParser.java:268]: A lot of magic values here, what do the 5 and 4 stand for here? If possible please delcare a constant that has a good name that explains the purpose of these values.

File: src/main/java/org/jabref/logic/importer/AuthorListParser.java [hunk 5]
```diff
@@ -209,6 +180,100 @@ public AuthorList parse(@NonNull String listOfNames) {
...
+        // handle useprefix=false
+        if ("false".equalsIgnoreCase(usePrefix)) {
+            namePrefix = null;
+        }
+
+        // abbreviate given name
+        String givenNameAbbreviated = abbreviateGivenName(givenName);
+
+        // create Author object
+        return Optional.of(new Author(givenName, givenNameAbbreviated, namePrefix, familyName, nameSuffix));
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/logic/importer/AuthorListParser.java:217]: This comment doesn't say anything that the code doesn't say already, it is useless, please remove it.

File: src/main/java/org/jabref/logic/importer/AuthorListParser.java [hunk 6]
```diff
@@ -238,6 +303,35 @@ private static String checkNamesCommaSeparated(String listOfNames) {
      * @return Preformatted author name; <CODE>Optional.empty()</CODE> if author name is empty.
      */
     private Optional<Author> getAuthor() {
+        int savedTokenStart = tokenStart;
+
+        while (tokenStart < original.length() && Character.isWhitespace(original.charAt(tokenStart))) {
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/logic/importer/AuthorListParser.java:308]: A comment above this block however would be helpful.

File: src/main/java/org/jabref/logic/importer/AuthorListParser.java [hunk 7]
```diff
@@ -238,6 +303,35 @@ private static String checkNamesCommaSeparated(String listOfNames) {
...
+            tokenStart++;
+        }
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/logic/importer/AuthorListParser.java:310]: A comment above this block however would be helpful.

**Comment by koppor** [comment]
Closing this issue due to inactivity :zzz:

**Pull request closed without merging**